README.md
# Platform V SDK

В этом репозитории содержатся SDK, с помощью которых вы сможете создавать подписанные запросы к сервисам Platform V.

Platform V – это набор программных продуктов, которые позволяют быстро создавать новые приложения и услуги с помощью развитой инфраструктуры Сбера.

Платформа включает сервисы:
* [Platform V Functions](https://developers.sber.ru/docs/ru/platform-v/functions/overview) — FaaS-решение для событийно-ориентированных бессерверных вычислений.
* [Platform V DataSpace](https://developers.sber.ru/docs/ru/platform-v/dataspace/overview) — отказоустойчивая и распределенная система для хранения и обработки данных в режиме онлайн, которая хранит данные в оперативной памяти и имеет всю функциональность СУБД.

О том как создавать подписанные запросы к сервисам Platform V, читайте в [документации](https://developers.sber.ru/docs/ru/platform-v/functions/calling-created-function#вызов-через-c-url).


java-sdk-3.1.1/changelog
3.0.6
add okhttp demo

3.0.5
add okhttp client for android development

3.0.2
fixed bugs

3.0.0
Simplified signature method and the format of signature header, not compatible with previous version

2.0.2
Repaired the problem of special character in URL, query, or header causes verification failure
Supported large file upload, and added examples


java-sdk-3.1.1/src/com/apig/sdk/demo/Main.java
//package name of the demo.
package com.apig.sdk.demo;

//Import the external dependencies.

import com.cloud.apigateway.sdk.utils.Client;
import com.cloud.apigateway.sdk.utils.Request;
import org.apache.http.Header;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpRequestBase;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

import java.io.IOException;

public class Main {

    public static void main(String[] args) {
        //Create a new request.
        Request request = new Request();
        try {
            //Set the request parameters.
            //AppKey, AppSecrect, Method and Url are required parameters.
            request.setKey("apigateway_sdk_demo_key");
            request.setSecret("apigateway_sdk_demo_secret");
            request.setMethod("GET");
            request.setUrl("https://30030113-3657-4fb6-a7ef-90764239b038.apigw.exampleRegion.com/app1?a=1");
            request.addHeader("Content-Type", "text/plain");
            //if it was published in other envs(except for Release),you need to add the information x-stage and the value is env's name
            //request.addHeader("x-stage", "publish_env_name");
            request.setBody("demo");
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        CloseableHttpClient client = null;
        try {
            //Sign the request.
            HttpRequestBase signedRequest = Client.sign(request);

            //Send the request.
            client = HttpClients.custom().build();
            HttpResponse response = client.execute(signedRequest);

            //Print the status line of the response.
            System.out.println(response.getStatusLine().toString());

            //Print the header fields of the response.
            Header[] resHeaders = response.getAllHeaders();
            for (Header h : resHeaders) {
                System.out.println(h.getName() + ":" + h.getValue());
            }

            //Print the body of the response.
            HttpEntity resEntity = response.getEntity();
            if (resEntity != null) {
                System.out.println(System.getProperty("line.separator") + EntityUtils.toString(resEntity, "UTF-8"));
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                if (client != null) {
                    client.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

}

java-sdk-3.1.1/src/com/apig/sdk/demo/OkHttpDemo.java
//package name of the demo.
package com.apig.sdk.demo;
import com.cloud.apigateway.sdk.utils.Client;
import com.cloud.apigateway.sdk.utils.Request;
import okhttp3.Headers;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import okhttp3.ResponseBody;
public class OkHttpDemo {

    public static void main(String[] args) {
        //Create a new request.
        Request request = new Request();
        try {
            request.setKey("apigateway_sdk_demo_key");
            request.setSecret("apigateway_sdk_demo_secret");
            request.setMethod("GET");
            request.setUrl("https://30030113-3657-4fb6-a7ef-90764239b038.apigw.exampleRegion.com/app1?name=value");
            request.addHeader("Content-Type", "text/plain");
            request.setBody("demo");
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
        try {
            //Sign the request.
            okhttp3.Request signedRequest = Client.signOkhttp(request);

            //Send the request.
            OkHttpClient client = new OkHttpClient.Builder().build();
            Response response = client.newCall(signedRequest).execute();

            //Print the status line of the response.
            System.out.println("status:" + response.code());

            //Print the header fields of the response.
            Headers resHeaders = response.headers();
            for (String h : resHeaders.names()) {
                System.out.println(h + ":" + resHeaders.get(h));
            }

            //Print the body of the response.
            ResponseBody resEntity = response.body();
            System.out.println("\n" + resEntity.string());

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}


java-sdk-3.1.1/src/com/apig/sdk/demo/WebSocketDemo.java
//package name of the demo.
package com.apig.sdk.demo;

import com.cloud.apigateway.sdk.utils.Client;
import com.cloud.apigateway.sdk.utils.Request;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;

public class WebSocketDemo {

    public static void main(String[] args) {
        //Create a new request.
        Request request = new Request();
        try {
            request.setKey("apigateway_sdk_demo_key");
            request.setSecret("apigateway_sdk_demo_secret");
            request.setMethod("GET");
            request.setUrl("https://30030113-3657-4fb6-a7ef-90764239b038.apigw.exampleRegion.com/ws?name=value");
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
        try {
            //Sign the request.
            okhttp3.Request signedRequest = Client.signOkhttp(request);

            //Send the request.
            OkHttpClient client = new OkHttpClient.Builder().build();
            WebSocketListener webSocketListener = new WebSocketListener() {
                @Override
                public void onMessage(WebSocket webSocket, String text) {
                    System.out.println("receive: " + text);
                }

                @Override
                public void onFailure(WebSocket webSocket, Throwable t, Response response) {
                    System.out.println("fail   : " + t.getMessage());
                }
            };
            WebSocket webSocket = client.newWebSocket(signedRequest, webSocketListener);
            for (int i = 0; i < 10; i++) {
                String msg = "hello," + System.currentTimeMillis();
                System.out.println("send   : " + msg);
                webSocket.send(msg);
                Thread.sleep(50);
            }
            webSocket.close(1000, null);
            client.dispatcher().executorService().shutdown();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}


java-sdk-3.1.1/src/com/apig/sdk/demo/LargeFileUploadDemo.java
//package name of the demo.
package com.apig.sdk.demo;

import com.cloud.apigateway.sdk.utils.Request;
import com.cloud.sdk.auth.vo.SignResult;
import com.cloud.sdk.util.BinaryUtils;
import com.cloud.sdk.util.SignUtils;

import javax.net.ssl.HttpsURLConnection;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.MessageDigest;
import java.util.Set;

public class LargeFileUploadDemo {
    private static final String HTTPS = "https";

    static String calcSha256Hex(String fileName) throws Exception {
        byte[] buffer = new byte[8192];
        int count;
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        BufferedInputStream bis = new BufferedInputStream(new FileInputStream(fileName));
        while ((count = bis.read(buffer)) > 0) {
            digest.update(buffer, 0, count);
        }
        byte[] hash = digest.digest();
        return BinaryUtils.toHex(hash);
    }

    public static void main(String[] args) {
        String fname = "example.rar";
        // Create a new request.
        Request request = new Request();
        try {
            // Set the request parameters.
            // AppKey, AppSecrect, Method and Url are required parameters.
            request.setKey("apigateway_sdk_demo_key");
            request.setSecret("apigateway_sdk_demo_secret");
            request.setMethod("POST");
            request.setUrl("https://30030113-3657-4fb6-a7ef-90764239b038.apigw.exampleRegion.com/app1?name=value");
            request.addHeader("Content-Type", "plain/text");
            String fileHash = calcSha256Hex(fname);
            request.addHeader("x-sdk-content-sha256", fileHash);
            System.out.println("x-sdk-content-sha256:" + fileHash);
            // if it was published in other envs(except for Release),you need to add the
            // information x-stage and the value is env's name
            // request.addHeader("x-stage", "publish_env_name");
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        URL uUrl = null;
        HttpURLConnection conn = null;
        OutputStream out = null;
        BufferedReader in = null;
        try {
            // Sign the request.
            SignResult signRet = SignUtils.sign(request);

            // initial connection
            uUrl = signRet.getUrl();
            if (uUrl.getProtocol().toLowerCase().equals(HTTPS)) {
                HttpsURLConnection httpsconn = (HttpsURLConnection) uUrl.openConnection();
                conn = httpsconn;
            } else {
                conn = (HttpURLConnection) uUrl.openConnection();
            }
            conn.setRequestMethod(request.getMethod().name());
            conn.setDoOutput(true);
            conn.setDoInput(true);

            // Send the request.
            if (signRet.getHeaders().size() > 0) {
                Set<String> headerSet = signRet.getHeaders().keySet();
                for (String key : headerSet) {
                    String value = signRet.getHeaders().get(key);
                    conn.setRequestProperty(key, value);
                }
            }
            // Send a file.
            setFileConent(fname, conn.getOutputStream());

            // Print the status line of the response.
            InputStream is = null;
            if (conn.getResponseCode() > 400) {
                is = conn.getErrorStream();
            } else {
                is = conn.getInputStream();
            }

            if (null != is) {
                StringBuilder result = new StringBuilder();
                in = new BufferedReader(new InputStreamReader(is, "UTF-8"));
                if (in != null) {
                    String line = "";
                    while ((line = in.readLine()) != null) {
                        result.append(line);
                    }
                }
                System.out.println(result.toString());
            }


        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                if (out != null) {
                    out.close();
                }
                if (in != null) {
                    in.close();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            if (conn != null) {
                conn.disconnect();
            }
        }
    }

    public static void setFileConent(String filePath, OutputStream out) throws IOException, FileNotFoundException {
        File file = new File(filePath);
        int count = 0;
        byte[] buffer = new byte[1024];
        BufferedInputStream bis = new BufferedInputStream(new FileInputStream(file));
        while ((count = bis.read(buffer)) > 0) {
            out.write(buffer, 0, count);
        }
        bis.close();
    }
}